book/docs/lectures/Lecture4.md

```diff
@@ -0,0 +1,243 @@
+Slot 00001: SlotAdd Slot 2
+Final balances
+Wallet 1:
+    {, ""}: 100000000
+Wallet 2:
+    {, ""}: 100000000
+Wallet 3:
+    {, ""}: 100000000
+Wallet 4:
+    {, ""}: 100000000
+Wallet 5:
+    {, ""}: 100000000
+Wallet 6:
+    {, ""}: 100000000
+Wallet 7:
+    {, ""}: 100000000
+Wallet 8:
+    {, ""}: 100000000
+Wallet 9:
+    {, ""}: 100000000
+Wallet 10:
+    {, ""}: 100000000
+```
+where, as we see, the message is no longer an error but a log from the contract that informs us about the error. As the error data type of `myContract2` is of type `Void` we can be sure it was handled.
+
+## The Schema parameter: s
+We can define a custom set of contract actions by adding this actions to the `BlockchainActions` type. For example, let us say we want to add and endpoint called 'foo'. We just need to give a pseudonym to the set of action data types like this:
+```haskell
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+-- ^ Add extensions on the top of the module file
+
+type MySchema = BlockchainActions .\/ Endpoint "foo" Int
+```
+In the last line we define the type of the set of actions and we call it `MySchema`. Then we use the operator `.\/`, which acts on types, not on values, to "add" the endpoints that we want, in this case the `foo` endpoint. The first argument to `Endpoint` is a type level string which represents the name of the endpoint, and the second argument is the parameter type (which type of value this endpoint takes).
+
+Once we have define the endpoint we can take the action defined by it using the trace emulator. First, we define our contract:
+```haskell
+myContract :: Contract () MySchema Text ()
+myContract = do
+    n <- endpoint @"foo"
+    Contract.logInfo n
+```
+This contract just waits for some wallet to call the `"foo"` endpoint with some `Int` value and then logs it to the console. Then, we define the trace of the simulation, where now we can use the endpoint we have just defined:
+```haskell
+myTrace :: EmulatorTrace ()
+myTrace = do
+    h <- activateContractWallet (Wallet 1) myContract
+    callEndpoint @"foo" h 42
+```
+Note that now we are interested in calling the endpoint, for which it is necessary to use the handler. We do not use anymore the `Void $ activateCon...` but we reference the function ouput with some variable, `h` (for handler) in this case. Finally we define the test function that runs this trace:
+```haskell
+test :: IO ()
+test = runEmulatorTraceIO myTrace
+```
+
+## The Writer parameter: w
+This type parameter can not be of any type but it must be an instance of the type class `Monoid`. An example of data type which is an instance of this class is `List`. This parameter of the Contract monad is essential because it allows us to bring information back from the contract to the trace and also to the _PAB_, the Plutus Application Backend. We will be able to pass info back from the contract running in the wallet to the outside world. Let us see an example:
```

Comment:
```suggestion
This type parameter can not be of any type but an instance of the type class `Monoid`. An example of data type which is an instance of this class is `List`. This parameter of the Contract monad is essential because it allows us to bring information back from the contract to the trace and also to the _PAB_, the Plutus Application Backend. We will be able to pass info back from the contract running in the wallet to the outside world. Let us see an example:
```